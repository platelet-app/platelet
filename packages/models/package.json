{
    "name": "@platelet-app/models",
    "version": "1.0.1",
    "description": "DataStore models for the platelet.app project",
    "homepage": "https://platelet.app",
    "bugs": {
        "url": "https://github.com/platelet-app/platelet/issues"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/platelet-app/platelet.git"
    },
    "license": "Apache-2.0",
    "author": "Theo Cranmore <theo@platelet.app>",
    "type": "module",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "build": "cd ../.. && amplify api gql-compile && amplify codegen models && cd packages/models && cp ../../models/* dist",
        "prepublishOnly": "npm version patch"
    }
}
