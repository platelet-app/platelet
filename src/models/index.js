// @ts-check
import { initSchema } from '@aws-amplify/datastore';
import { schema } from './schema';

const Role = {
  "USER": "USER",
  "COORDINATOR": "COORDINATOR",
  "RIDER": "RIDER",
  "ADMIN": "ADMIN"
};

const CommentVisibility = {
  "EVERYONE": "EVERYONE",
  "ME": "ME"
};

const TaskStatus = {
  "NEW": "NEW",
  "ACTIVE": "ACTIVE",
  "PICKED_UP": "PICKED_UP",
  "DROPPED_OFF": "DROPPED_OFF",
  "CANCELLED": "CANCELLED",
  "REJECTED": "REJECTED",
  "ABANDONED": "ABANDONED",
  "COMPLETED": "COMPLETED"
};

const Priority = {
  "HIGH": "HIGH",
  "MEDIUM": "MEDIUM",
  "LOW": "LOW"
};

const DeliverableTypeIcon = {
  "BUG": "BUG",
  "CHILD": "CHILD",
  "DOCUMENT": "DOCUMENT",
  "EQUIPMENT": "EQUIPMENT",
  "OTHER": "OTHER"
};

const DeliverableUnit = {
  "NONE": "NONE",
  "LITER": "LITER",
  "MILLILITER": "MILLILITER",
  "GRAM": "GRAM",
  "ITEM": "ITEM",
  "BOX": "BOX"
};

const { User, PossibleRiderResponsibilities, RiderResponsibility, Comment, TaskAssignee, Task, Location, Handover, Deliverable, DeliverableType, VehicleAssignment, Vehicle, Tenant, Statistics, SendFeedback, AddressAndContactDetails, S3Object } = initSchema(schema);

export {
  User,
  PossibleRiderResponsibilities,
  RiderResponsibility,
  Comment,
  TaskAssignee,
  Task,
  Location,
  Handover,
  Deliverable,
  DeliverableType,
  VehicleAssignment,
  Vehicle,
  Tenant,
  Role,
  CommentVisibility,
  TaskStatus,
  Priority,
  DeliverableTypeIcon,
  DeliverableUnit,
  Statistics,
  SendFeedback,
  AddressAndContactDetails,
  S3Object
};