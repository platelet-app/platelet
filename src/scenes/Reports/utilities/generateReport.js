import { DataStore } from "aws-amplify";
import _ from "lodash";
import moment from "moment";
import { commentVisibility } from "../../../apiConsts";
import * as models from "../../../models";
import { writeToString } from "@fast-csv/format";

const taskFields = {
    createdAt: "",
    isRiderUsingOwnVehicle: "",
    timeOfCall: "",
    priority: "",
    status: "",
    riderResponsibility: "",
    timePickedUp: "",
    timeDroppedOff: "",
    timeRiderHome: "",
};

const locationFields = {
    ward: "",
    line1: "",
    line2: "",
    line3: "",
    town: "",
    county: "",
    state: "",
    country: "",
    postcode: "",
};

const itemFields = {
    createdAt: "",
    label: "",
    count: "",
    unit: "",
};

const requesterContactFields = {
    name: "",
    telephoneNumber: "",
};

const assigneeFields = {
    createdAt: "",
    displayName: "",
    name: "",
    role: "",
};

const commentFields = {
    createdAt: "",
    author: "",
    body: "",
};

function generateHeader(fields, prefix = "") {
    let keys = Object.keys(fields);
    if (prefix) {
        return keys.map((k) => `${prefix}_${k}`);
    } else {
        return keys;
    }
}

function generateCountedHeader(count, fields, prefix) {
    return _.range(count)
        .map((i) => {
            return Object.keys(fields).map((k) => `${prefix}_${i}_${k}`);
        })
        .flat();
}

async function generateCSV(data) {
    const rows = [];
    const headers = [];
    const commentsCount = data.reduce((acc, task) => {
        if (task.comments && task.comments.length > acc)
            return task.comments.length;
        else return acc;
    }, 0);
    const itemOffsetCount =
        [
            ...Object.keys(taskFields),
            ...Object.keys(locationFields),
            ...Object.keys(locationFields),
            ...Object.keys(requesterContactFields),
        ].length +
        commentsCount * Object.keys(commentFields).length;

    const itemsCount = data.reduce((acc, task) => {
        if (task.items && task.items.length > acc) return task.items.length;
        else return acc;
    }, 0);

    const assigneeOffsetCount =
        itemOffsetCount + itemsCount * Object.keys(itemFields).length;

    const assigneesCount = data.reduce((acc, task) => {
        if (task.assignees && task.assignees.length > acc)
            return task.assignees.length;
        else return acc;
    }, 0);

    headers.push(generateHeader(taskFields));
    headers.push(generateHeader(locationFields, "pickUpLocation"));
    headers.push(generateHeader(locationFields, "dropOffLocation"));
    headers.push(generateHeader(requesterContactFields, "requesterContact"));
    if (commentsCount > 0) {
        headers.push(
            generateCountedHeader(commentsCount, commentFields, "comment")
        );
    }
    if (itemsCount > 0) {
        headers.push(generateCountedHeader(itemsCount, itemFields, "item"));
    }
    if (assigneesCount > 0) {
        headers.push(
            generateCountedHeader(assigneesCount, assigneeFields, "assignee")
        );
    }
    console.log(headers);
    rows.push(headers.flat());
    console.log(rows);
    data.forEach((item) => {
        let row = [];
        const {
            pickUpLocation,
            dropOffLocation,
            requesterContact,
            comments,
            items,
            assignees,
            createdBy,
            ...rest
        } = item;
        Object.keys(taskFields).forEach((key) => {
            row.push(rest[key]);
        });
        if (pickUpLocation)
            Object.keys(locationFields).forEach((key) => {
                row.push(pickUpLocation[key] || locationFields[key]);
            });
        else
            Object.values(locationFields).forEach((value) => {
                row.push(value);
            });

        if (dropOffLocation)
            Object.keys(locationFields).forEach((key) => {
                row.push(dropOffLocation[key] || locationFields[key]);
            });
        else
            Object.values(locationFields).forEach((value) => {
                row.push(value);
            });
        if (requesterContact)
            Object.keys(requesterContactFields).forEach((key) => {
                row.push(requesterContact[key]);
            });
        else
            Object.values(requesterContactFields).forEach((value) => {
                row.push(value);
            });
        if (comments && comments.length > 0) {
            comments.forEach((comment) => {
                Object.keys(commentFields).forEach((key) => {
                    if (key === "author") {
                        row.push(
                            comment[key].displayName || commentFields[key]
                        );
                    } else {
                        row.push(comment[key] || commentFields[key]);
                    }
                });
            });
        }

        if (items && items.length > 0) {
            if (row.length !== itemOffsetCount) {
                _.range(itemOffsetCount - row.length).forEach(() => {
                    row.push("");
                });
            }
            items.forEach((item) => {
                Object.keys(itemFields).forEach((key) => {
                    if (key === "label") {
                        row.push(item.deliverableType.label || itemFields[key]);
                    } else {
                        row.push(item[key] || itemFields[key]);
                    }
                });
            });
        }

        if (assignees && assignees.length > 0) {
            if (row.length !== assigneeOffsetCount) {
                _.range(assigneeOffsetCount - row.length).forEach(() => {
                    row.push("");
                });
            }
            assignees.forEach((item) => {
                Object.keys(assigneeFields).forEach((key) => {
                    if (["displayName", "name"].includes(key)) {
                        row.push(item.assignee[key] || assigneeFields[key]);
                    } else {
                        row.push(item[key] || assigneeFields[key]);
                    }
                });
            });
        }

        rows.push(row);
    });
    return await writeToString(rows);
}

export default async function generateReport(userId, role, days) {
    const timeStamp = moment
        .utc()
        .subtract(days.toString(), "days")
        .toISOString();
    let finalTasks = [];
    const assignments = await DataStore.query(models.TaskAssignee);
    if (role !== "ALL") {
        const filteredAssignments = assignments.filter(
            (assignment) =>
                assignment.task &&
                assignment.role === role &&
                assignment.assignee &&
                assignment.assignee.id === userId &&
                assignment.task
        );
        const taskIds = filteredAssignments.map(
            (assignment) => assignment.task.id
        );
        if (taskIds.length === 0) {
            finalTasks = [];
        } else {
            finalTasks = await DataStore.query(models.Task, (task) =>
                task
                    .or((task) =>
                        task
                            .createdAt("eq", undefined)
                            .createdAt("gt", timeStamp)
                    )
                    .or((task) =>
                        taskIds.reduce((task, id) => task.id("eq", id), task)
                    )
            );
        }
    } else if (role === "ALL") {
        finalTasks = await DataStore.query(models.Task, (task) =>
            task.or((task) =>
                task.createdAt("eq", undefined).createdAt("gt", timeStamp)
            )
        );
    }
    const deliverables = await DataStore.query(models.Deliverable);
    const commentsAll = await DataStore.query(models.Comment, (c) =>
        c.visibility("eq", commentVisibility.everyone)
    );
    finalTasks = await Promise.all(
        finalTasks.map(async (t) => {
            const isRiderUsingOwnVehicle = !!t.isRiderUsingOwnVehicle;
            const comments = commentsAll.filter((c) => c.parentId === t.id);
            const items = deliverables.filter(
                (d) => d.task && d.task.id === t.id
            );
            const assignees = assignments.filter(
                (assignment) => assignment.task && assignment.task.id === t.id
            );
            return { ...t, comments, items, assignees, isRiderUsingOwnVehicle };
        })
    );
    return generateCSV(finalTasks);
}

// callback code that may be useful
/*let count = 0;
    finalTasks = await Promise.all(
        finalTasks.map((t) =>
            new Promise((resolve, reject) => {
                const comments = commentsAll.filter((c) => c.parentId === t.id);
                const items = deliverables.filter(
                    (d) => d.task && d.task.id === t.id
                );
                const assignees = assignments.filter(
                    (assignment) =>
                        assignment.task && assignment.task.id === t.id
                );
                resolve({ ...t, comments, items, assignees });
            })
                .then((result) => {
                    count++;
                    if (progressCallback) {
                        progressCallback((count * 100) / finalTasks.length);
                    }
                    return result;
                })
                .catch((err) => {
                    console.log(err);
                    throw err;
                })
        )
    );
    */
