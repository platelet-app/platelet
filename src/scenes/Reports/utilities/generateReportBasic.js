import { DataStore } from "aws-amplify";
import _ from "lodash";
import moment from "moment";
import { commentVisibility } from "../../../apiConsts";
import * as models from "../../../models";
import { writeToString } from "@fast-csv/format";

const taskFields = {
    timeOfCall: "",
    riderResponsibility: "",
    isRiderUsingOwnVehicle: "",
    priority: "",
    status: "",
    timePickedUp: "",
    timeDroppedOff: "",
    timeRiderHome: "",
};

const locationFields = {
    ward: "",
    line1: "",
    town: "",
    postcode: "",
};

const itemFields = {
    label: "",
    count: "",
};

const requesterContactFields = {
    name: "",
    telephoneNumber: "",
};

const assigneeFields = {
    name: "",
    role: "",
};

const commentFields = {
    author: "",
    body: "",
};

function generateHeader(fields, prefix = "") {
    let keys = Object.keys(fields);
    if (prefix) {
        return keys.map((k) => `${prefix}_${k}`);
    } else {
        return keys;
    }
}

function generateCountedHeader(count, fields, prefix) {
    return _.range(count)
        .map((i) => {
            return Object.keys(fields).map((k) => `${prefix}_${i}_${k}`);
        })
        .flat();
}

async function generateCSV(data) {
    const rows = [];
    const headers = [];
    const assigneesCount = data.reduce((acc, task) => {
        if (task.assignees && task.assignees.length > acc)
            return task.assignees.length;
        else return acc;
    }, 0);
    const itemOffsetCount =
        [
            ...Object.keys(taskFields),
            ...Object.keys(locationFields),
            ...Object.keys(locationFields),
            ...Object.keys(requesterContactFields),
        ].length +
        assigneesCount * Object.keys(assigneeFields).length;

    const itemsCount = data.reduce((acc, task) => {
        if (task.items && task.items.length > acc) return task.items.length;
        else return acc;
    }, 0);

    const commentsOffsetCount =
        itemOffsetCount + itemsCount * Object.keys(itemFields).length;

    const commentsCount = data.reduce((acc, task) => {
        if (task.comments && task.comments.length > acc)
            return task.comments.length;
        else return acc;
    }, 0);

    headers.push(generateHeader(taskFields));
    headers.push(generateHeader(requesterContactFields, "requesterContact"));
    headers.push(generateHeader(locationFields, "pickUpLocation"));
    headers.push(generateHeader(locationFields, "dropOffLocation"));
    if (assigneesCount > 0) {
        headers.push(
            generateCountedHeader(assigneesCount, assigneeFields, "assignee")
        );
    }
    if (itemsCount > 0) {
        headers.push(generateCountedHeader(itemsCount, itemFields, "item"));
    }
    if (commentsCount > 0) {
        headers.push(
            generateCountedHeader(commentsCount, commentFields, "comment")
        );
    }
    console.log(headers);
    rows.push(headers.flat());
    console.log(rows);
    data.forEach((item) => {
        let row = [];
        const {
            pickUpLocation,
            dropOffLocation,
            requesterContact,
            comments,
            items,
            assignees,
            createdBy,
            ...rest
        } = item;
        Object.keys(taskFields).forEach((key) => {
            row.push(rest[key]);
        });
        if (requesterContact)
            Object.keys(requesterContactFields).forEach((key) => {
                row.push(requesterContact[key]);
            });
        else
            Object.values(requesterContactFields).forEach((value) => {
                row.push(value);
            });
        if (pickUpLocation)
            Object.keys(locationFields).forEach((key) => {
                row.push(pickUpLocation[key] || locationFields[key]);
            });
        else
            Object.values(locationFields).forEach((value) => {
                row.push(value);
            });

        if (dropOffLocation)
            Object.keys(locationFields).forEach((key) => {
                row.push(dropOffLocation[key] || locationFields[key]);
            });
        else
            Object.values(locationFields).forEach((value) => {
                row.push(value);
            });

        if (assignees && assignees.length > 0) {
            assignees.forEach((item) => {
                Object.keys(assigneeFields).forEach((key) => {
                    if (["displayName", "name"].includes(key)) {
                        row.push(item.assignee[key] || assigneeFields[key]);
                    } else {
                        row.push(item[key] || assigneeFields[key]);
                    }
                });
            });
        }

        if (items && items.length > 0) {
            if (row.length !== itemOffsetCount) {
                _.range(itemOffsetCount - row.length).forEach(() => {
                    row.push("");
                });
            }
            items.forEach((item) => {
                Object.keys(itemFields).forEach((key) => {
                    if (key === "label") {
                        row.push(item.deliverableType.label || itemFields[key]);
                    } else {
                        row.push(item[key] || itemFields[key]);
                    }
                });
            });
        }

        if (comments && comments.length > 0) {
            if (row.length !== commentsOffsetCount) {
                _.range(commentsOffsetCount - row.length).forEach(() => {
                    row.push("");
                });
            }
            comments.forEach((comment) => {
                Object.keys(commentFields).forEach((key) => {
                    if (key === "author") {
                        row.push(
                            comment[key].displayName || commentFields[key]
                        );
                    } else {
                        row.push(comment[key] || commentFields[key]);
                    }
                });
            });
        }

        rows.push(row);
    });
    return await writeToString(rows);
}

export default async function generateReportBasic(userId, role, days) {
    const timeStamp = moment
        .utc()
        .subtract(days.toString(), "days")
        .toISOString();
    let finalTasks = [];
    const assignments = await DataStore.query(models.TaskAssignee);
    if (role !== "ALL") {
        const filteredAssignments = assignments.filter(
            (assignment) =>
                assignment.task &&
                assignment.role === role &&
                assignment.assignee &&
                assignment.assignee.id === userId &&
                assignment.task
        );
        const taskIds = filteredAssignments.map(
            (assignment) => assignment.task.id
        );
        if (taskIds.length === 0) {
            finalTasks = [];
        } else {
            finalTasks = await DataStore.query(models.Task, (task) =>
                task
                    .or((task) =>
                        task
                            .createdAt("eq", undefined)
                            .createdAt("gt", timeStamp)
                    )
                    .or((task) =>
                        taskIds.reduce((task, id) => task.id("eq", id), task)
                    )
            );
        }
    } else if (role === "ALL") {
        finalTasks = await DataStore.query(models.Task, (task) =>
            task.or((task) =>
                task.createdAt("eq", undefined).createdAt("gt", timeStamp)
            )
        );
    }
    const deliverables = await DataStore.query(models.Deliverable);
    const commentsAll = await DataStore.query(models.Comment, (c) =>
        c.visibility("eq", commentVisibility.everyone)
    );
    finalTasks = await Promise.all(
        finalTasks.map(async (t) => {
            const isRiderUsingOwnVehicleBool = !!t.isRiderUsingOwnVehicle;
            const isRiderUsingOwnVehicle = isRiderUsingOwnVehicleBool
                ? "TRUE"
                : "FALSE";

            const comments = commentsAll.filter((c) => c.parentId === t.id);
            const items = deliverables.filter(
                (d) => d.task && d.task.id === t.id
            );
            const assignees = assignments.filter(
                (assignment) => assignment.task && assignment.task.id === t.id
            );
            return { ...t, comments, items, assignees, isRiderUsingOwnVehicle };
        })
    );
    return generateCSV(finalTasks);
}

// callback code that may be useful
/*let count = 0;
    finalTasks = await Promise.all(
        finalTasks.map((t) =>
            new Promise((resolve, reject) => {
                const comments = commentsAll.filter((c) => c.parentId === t.id);
                const items = deliverables.filter(
                    (d) => d.task && d.task.id === t.id
                );
                const assignees = assignments.filter(
                    (assignment) =>
                        assignment.task && assignment.task.id === t.id
                );
                resolve({ ...t, comments, items, assignees });
            })
                .then((result) => {
                    count++;
                    if (progressCallback) {
                        progressCallback((count * 100) / finalTasks.length);
                    }
                    return result;
                })
                .catch((err) => {
                    console.log(err);
                    throw err;
                })
        )
    );
    */
