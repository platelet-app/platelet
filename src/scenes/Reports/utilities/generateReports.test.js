import { DataStore } from "aws-amplify";
import {
    commentVisibility,
    deliverableUnits,
    priorities,
    tasksStatus,
    userRoles,
} from "../../../apiConsts";
import * as models from "../../../models";
import generateReport from "./generateReport";
import generateReportBasic from "./generateReportBasic";

const pickUpLocationData = {
    listed: 1,
    name: "pickup location",
    ward: "pickup ward1",
    line1: "pickup line1",
    line2: "pickup line2",
    line3: "pickup line3",
    town: "pickup town",
    county: "pickup county",
    postcode: "pickup postcode",
    state: "pickup state",
    country: "pickup country",
};
const dropOffLocationData = {
    listed: 1,
    name: "dropoff location",
    ward: "dropoff ward",
    line1: "dropoff line1",
    line2: "dropoff line2",
    line3: "dropoff line3",
    town: "dropoff town",
    county: "dropoff county",
    postcode: "dropoff postcode",
    state: "dropoff state",
    country: "dropoff country",
};

const locReducer = (acc, [key, value]) => {
    if (key === "listed") return acc;
    acc[key] = value + " 2";
    return acc;
};

describe("generateReports", () => {
    const RealDate = Date;
    const isoDate = "2021-11-29T23:24:58.987Z";

    function mockDate() {
        global.Date = class extends RealDate {
            constructor() {
                super();
                return new RealDate(isoDate);
            }
        };
    }

    afterEach(async () => {
        global.Date = RealDate;
    });

    beforeEach(async () => {
        jest.restoreAllMocks();
        mockDate();
    });

    test("generate a report", async () => {
        const rider1 = new models.User({
            displayName: "Rider 1",
            name: "Rider 1 name",
            roles: [userRoles.rider],
        });
        const rider2 = new models.User({
            displayName: "Rider 2",
            name: "Rider 2 name",
            roles: [userRoles.rider],
        });
        const coord = new models.User({
            displayName: "Coordinator",
            name: "Coordinator name",
            roles: [userRoles.coordinator],
        });
        const whoami = new models.User({
            displayName: "Whoami",
            name: "Whoami name",
            roles: [userRoles.coordinator],
        });

        const pickUpLocation1 = await DataStore.save(
            new models.Location(pickUpLocationData)
        );
        const pickUpLocationData2 = Object.entries(pickUpLocationData).reduce(
            locReducer,
            {}
        );
        const pickUpLocation2 = await DataStore.save(
            new models.Location(pickUpLocationData2)
        );
        const dropOffLocationData2 = Object.entries(dropOffLocationData).reduce(
            locReducer,
            {}
        );

        const dropOffLocation1 = await DataStore.save(
            new models.Location(dropOffLocationData)
        );
        const dropOffLocation2 = await DataStore.save(
            new models.Location({ ...dropOffLocationData2, listed: 0 })
        );
        const task1 = await DataStore.save(
            new models.Task({
                status: tasksStatus.new,
                timeOfCall: new Date().toISOString(),
                requesterContact: {
                    name: "some person",
                    telephoneNumber: "0121212121",
                },
                priority: priorities.medium,
                isRiderUsingOwnVehicle: 0,
                pickUpLocation: pickUpLocation1,
                dropOffLocation: dropOffLocation1,
                riderResponsibility: "yay rider role",
                timePickedUp: new Date().toISOString(),
                timeDroppedOff: new Date().toISOString(),
                timeRiderHome: new Date().toISOString(),
            })
        );
        const task2 = await DataStore.save(
            new models.Task({
                status: tasksStatus.active,
                timeOfCall: new Date().toISOString(),
                requesterContact: {
                    name: "some name",
                    telephoneNumber: "01234567890",
                },
                priority: priorities.medium,
                isRiderUsingOwnVehicle: 1,
                pickUpLocation: pickUpLocation2,
                dropOffLocation: dropOffLocation2,
                riderResponsibility: "some rider role",
                timePickedUp: new Date().toISOString(),
                timeDroppedOff: new Date().toISOString(),
                timeRiderHome: new Date().toISOString(),
            })
        );
        const [coordAssignee1, coordAssignee2] = await Promise.all(
            [task1, task2].map((t) =>
                DataStore.save(
                    new models.TaskAssignee({
                        task: t,
                        assignee: whoami,
                        role: userRoles.coordinator,
                    })
                )
            )
        );
        const riderAssignee1 = await DataStore.save(
            new models.TaskAssignee({
                task: task1,
                assignee: rider1,
                role: userRoles.rider,
            })
        );
        const riderAssignee2 = await DataStore.save(
            new models.TaskAssignee({
                task: task2,
                assignee: rider2,
                role: userRoles.rider,
            })
        );
        const coordAssignee = await DataStore.save(
            new models.TaskAssignee({
                task: task1,
                assignee: coord,
                role: userRoles.coordinator,
            })
        );
        const deli1 = await DataStore.save(
            new models.DeliverableType({ label: "deliverable 1" })
        );
        const deli2 = await DataStore.save(
            new models.DeliverableType({ label: "deliverable 2" })
        );
        const deli3 = await DataStore.save(
            new models.DeliverableType({ label: "deliverable 3" })
        );
        const item1 = await DataStore.save(
            new models.Deliverable({
                task: task1,
                deliverableType: deli1,
                unit: deliverableUnits.box,
                count: 5,
            })
        );
        const item2 = await DataStore.save(
            new models.Deliverable({
                task: task2,
                deliverableType: deli2,
                unit: deliverableUnits.item,
                count: 6,
            })
        );
        const item3 = await DataStore.save(
            new models.Deliverable({
                task: task1,
                deliverableType: deli3,
                unit: deliverableUnits.liter,
                count: 7,
            })
        );
        const [comment1, comment2] = await Promise.all(
            ["body 1", "body 2"].map((body) =>
                DataStore.save(
                    new models.Comment({
                        body,
                        parentId: task1.id,
                        author: rider1,
                        visibility: commentVisibility.everyone,
                    })
                )
            )
        );

        const result = await generateReport(
            whoami.id,
            userRoles.coordinator,
            3
        );
        expect(result).toMatchSnapshot();
        const resultBasic = await generateReportBasic(
            whoami.id,
            userRoles.coordinator,
            3
        );
        expect(resultBasic).toMatchSnapshot();
    });
});
