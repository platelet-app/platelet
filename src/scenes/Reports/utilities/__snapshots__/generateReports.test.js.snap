// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateReports generate a report with ALL tasks 2`] = `
"createdAt,timeOfCall,riderResponsibility,isRiderUsingOwnVehicle,priority,status,timePickedUp,timeDroppedOff,timeRiderHome,timeRejected,timeCancelled,requesterContact_name,requesterContact_telephoneNumber,pickUpLocation_ward,pickUpLocation_line1,pickUpLocation_town,pickUpLocation_postcode,pickUpLocation_listed,dropOffLocation_ward,dropOffLocation_line1,dropOffLocation_town,dropOffLocation_postcode,dropOffLocation_listed,assignee_0_name,assignee_0_role,assignee_1_name,assignee_1_role,assignee_2_name,assignee_2_role,item_0_label,item_0_count,item_1_label,item_1_count,comment_0_author,comment_0_body,comment_1_author,comment_1_body
2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,yay rider role,FALSE,MEDIUM,NEW,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,,2021-11-29T23:24:58.987Z,some person,0121212121,pickup ward1,pickup line1,pickup town,pickup postcode,1,dropoff ward,dropoff line1,dropoff town,dropoff postcode,1,Coordinator name,COORDINATOR,Whoami name,COORDINATOR,Rider 1 name,RIDER,deliverable 1,5,deliverable 2,6,Rider 1,body 1,Rider 1,body 2
2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,yay rider role,FALSE,MEDIUM,ACTIVE,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,,some name,01234567890,pickup ward1,pickup line1,pickup town,pickup postcode,1,dropoff ward,dropoff line1,dropoff town,dropoff postcode,1,,,Rider 2 name,RIDER,,,deliverable 3,7"
`;

exports[`generateReports generate a report with a role 1`] = `
"createdAt,timeOfCall,riderResponsibility,isRiderUsingOwnVehicle,priority,status,timePickedUp,timeDroppedOff,timeRiderHome,timeRejected,timeCancelled,requesterContact_name,requesterContact_telephoneNumber,pickUpLocation_ward,pickUpLocation_line1,pickUpLocation_town,pickUpLocation_postcode,pickUpLocation_listed,dropOffLocation_ward,dropOffLocation_line1,dropOffLocation_town,dropOffLocation_postcode,dropOffLocation_listed,assignee_0_name,assignee_0_role,assignee_1_name,assignee_1_role,assignee_2_name,assignee_2_role,item_0_label,item_0_count,item_1_label,item_1_count,comment_0_author,comment_0_body,comment_1_author,comment_1_body
,2021-11-29T23:24:58.987Z,yay rider role,FALSE,MEDIUM,NEW,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,,2021-11-29T23:24:58.987Z,some person,0121212121,pickup ward1,pickup line1,pickup town,pickup postcode,TRUE,dropoff ward,dropoff line1,dropoff town,dropoff postcode,TRUE,Whoami name,COORDINATOR,Rider 1 name,RIDER,Coordinator name,COORDINATOR,deliverable 1,5,deliverable 3,7,Rider 1,body 1,Rider 1,body 2
,2021-11-29T23:24:58.987Z,some rider role,TRUE,MEDIUM,ACTIVE,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,2021-11-29T23:24:58.987Z,,some name,01234567890,pickup ward1 2,pickup line1 2,pickup town 2,pickup postcode 2,FALSE,dropoff ward 2,dropoff line1 2,dropoff town 2,dropoff postcode 2,FALSE,Whoami name,COORDINATOR,Rider 2 name,RIDER,,,deliverable 2,6"
`;
